### PR TITLE
bug fixes in lib.distances

### PR DESCRIPTION
**Pull request by zemanj**
Fixes (partially) #2046. This is the fourth of a series of related PRs following PRs #2048, #2062, and #2070.

Changes made in this Pull Request:
- Ensured that none of the functions in `MDAnalysis.lib.distances` modifies its input coordinate arrays in-place. This fixes a bug I introduced in PR #2048 (see updated issue #2046).
- Allowed for empty coordinate array input (`shape=(0,3)`) in all `lib.distances` functions.
- Fixed failures / wrong distance output of `*capped_distance()` in several corner cases.
- Added corresponding tests ensuring non-modified input, correct functionality with empty input, and correct return types.
- Improved return type specifications in the docs.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by zemanj** [comment]
This is *not* ready for review. Since I have to add tests for empty inputs to satisfy the diff coverage, I'll do that for all functions right away.

**Comment by richardjgowers** [comment]
@zemanj how/where were coordinates being modified in place?

**Comment by zemanj** [comment]
@richardjgowers 
> @zemanj how/where were coordinates being modified in place?

This happened when I introduced the new @check_coords() decorator. For some functions, I falsely assumed that they either make copies of (or only read from) the input arrays internally. I fixed the problem by removing the `enforce_copy=False` decorator keyword in these cases. In order to guarantee unmodified input arrays, I added the class `TestInputUnchanged` to `lib/test_distances.py`.

File: package/MDAnalysis/lib/distances.py [hunk 1]
```diff
@@ -571,33 +573,36 @@ def _bruteforce_capped(reference, configuration, max_cutoff, min_cutoff=None,
...
         ``k``-th pair returned in `pairs` and gives the distance between the
         coordinates ``reference[pairs[k, 0]]`` and
         ``configuration[pairs[k, 1]]``.
     """
-    distances = distance_array(reference, configuration, box=box)
-    if min_cutoff is not None:
-        mask = np.where((distances <= max_cutoff) & (distances > min_cutoff))
-    else:
-        mask = np.where((distances <= max_cutoff))
+    pairs = np.empty((0, 2), dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (zemanj) [reply package/MDAnalysis/lib/distances.py]: Ok, I'll add some explanatory comments!

File: package/MDAnalysis/lib/distances.py [hunk 2]
```diff
@@ -1003,34 +1015,38 @@ def _pkdtree_capped_self(reference, max_cutoff, min_cutoff=None, box=None):
...
-        distance = calc_bonds(reference[refA], reference[refB], box=box)
-        if min_cutoff is not None:
-            mask = np.where(distance > min_cutoff)[0]
-            pairs, distance = pairs[mask], distance[mask]
-    return np.asarray(pairs), np.asarray(distance)
-
+    pairs = np.empty((0, 2), dtype=np.int64)
+    distances = np.empty((0,), dtype=np.float64)
+
+    if len(reference) > 1:
```
[the remark this answers is not in the document]
Comment (zemanj) [reply package/MDAnalysis/lib/distances.py]: Sure, will do!

**Comment by zemanj** [comment]
@richardjgowers I added the requested comments. Moreover, I removed the obsolete dtype conversion in the return statements of `capped_distance()` and `self_capped_distance()`. This has the advantage that the return type is now also tested for the individual methods (not really required IMHO, but nice to have).

**Comment by zemanj** [comment]
@richardjgowers Do you want to take another look or shall I merge?

**Comment by zemanj** [comment]
@richardjgowers Thank you for reviewing!
